### DISCUSSION TITLE
How to declare cronjob in Hono??

### DISCUSSION THREAD

**Discussion by dnguyenfs** [honojs/hono#1641] [Ideas]
How can I create cronjob in Hono guys?

**Comment by yusukebe** [comment]
I'm not sure what you meant by "cronjob", but Hono is a slim framework, almost like a router, so it doesn't have a cron function.

However, if you were referring to Cloudflare Triggers, you can write it as follows:

```ts
export default {
  fetch: app.fetch,
  async scheduled(event, env, ctx) {
    ctx.waitUntil(doSomeTaskOnASchedule())
  }
}
```

**Reply by r0ckerz** [reply]
Can you please provide more details here?
What are (event, env, ctx) types? Why is ctx.waitUntil used here instead of event.waitUntil?
Is it possible to pass down env.Bindings from the main app so that KVNamespace can be accessed in a scheduled task? What would be the best way to do that? Thanks!

**Reply by EyalRo** [reply]
I tried this, and it breaks the hono routes (getting error 500). It also fails to expose the Cloudflare /__scheduled

**Reply by exoup** [reply]
> Can you please provide more details here? What are (event, env, ctx) types? Why is ctx.waitUntil used here instead of event.waitUntil? Is it possible to pass down env.Bindings from the main app so that KVNamespace can be accessed in a scheduled task? What would be the best way to do that? Thanks!

@r0ckerz The example seems to come from the Cloudflare Workers docs [here](https://developers.cloudflare.com/workers/runtime-apis/handlers/scheduled/#syntax), which explains a bit about the various parameters.

**Comment by EyalRo** [comment]
I got this working nicely with something like this (in typescript): note that cronTask is just a function called on the cron schedule. You can use whatever works for you. I use the waitUntil method so I can guarantee to finish the cronTask function processing before the cloud function is terminated (all of this is running on Cloudflare workers):

```typescript
export default {
/** this part manages cronjobs */
  scheduled(
    event: ScheduledEvent,
    env: {
      DB?: any;
      SENDGRID_API_KEY: any;
    },
    ctx: ExecutionContext
  ) {
    const delayedProcessing = async () => {
      await cronTask(env);
    };
    ctx.waitUntil(delayedProcessing());
  },
/** this part manages regular REST */
  fetch(request: Request, env: Env, ctx: ExecutionContext) {
    return app.fetch(request, env, ctx);
  },
};
```

in plain JS this should look something like this (js code untested):
```javascript
export default {
  scheduled(event, env, ctx) {
    const delayedProcessing = async () => {
      await cronTask(env)
    }
    ctx.waitUntil(delayedProcessing())
  },
  fetch(request, env, ctx) {
    return app.fetch(request, env, ctx)
  }
}
```

**Reply by exoup** [reply]
Very helpful. Thanks!

**Reply by zinc1234596** [reply]
really thankful! This method works! ❤️

**Reply by kirtirajsinh** [reply]
> I got this working nicely with something like this (in typescript): note that cronTask is just a function called on the cron schedule. You can use whatever works for you. I use the waitUntil method so I can guarantee to finish the cronTask function processing before the cloud function is terminated (all of this is running on Cloudflare workers):
> 
> ```ts
> export default {
> /** this part manages cronjobs */
>   scheduled(
>     event: ScheduledEvent,
>     env: {
>       DB?: any;
>       SENDGRID_API_KEY: any;
>     },
>     ctx: ExecutionContext
>   ) {
>     const delayedProcessing = async () => {
>       await cronTask(env);
>     };
>     ctx.waitUntil(delayedProcessing());
>   },
> /** this part manages regular REST */
>   fetch(request: Request, env: Env, ctx: ExecutionContext) {
>     return app.fetch(request, env, ctx);
>   },
> };
> ```
> 
> in plain JS this should look something like this (js code untested):
> 
> ```js
> export default {
>   scheduled(event, env, ctx) {
>     const delayedProcessing = async () => {
>       await cronTask(env)
>     }
>     ctx.waitUntil(delayedProcessing())
>   },
>   fetch(request, env, ctx) {
>     return app.fetch(request, env, ctx)
>   }
> }
> ```

This is my code and I tried all way but the scheduled function is not being called. 

```
import { Hono } from "hono";
// import { handlePostResponse } from "./utils";
import { Bindings } from "./env.types";
import { handlePostResponse, replyWithARoast, roastthePost } from "./utils";
import { generateImage } from "./generate-image";
import { upload } from "../lib/upload";

export const app = new Hono<{ Bindings: Bindings }>();

app.get("/", async (c) => {

  return c.text("Hello buddha!");
});

app.post("/event", async (c) => {
  const body = await c.req.json();

  console.log(body, "original body");
  const response = await handlePostResponse(
    body,
    c.env
  );


  return c.json({
    data: response
  });
});

app.post("/generate-image", async (c) => {
  const body = await c.req.json();
  console.log(body, "original body");
  const svg = await generateImage(body.text, c.env);
  console.log(svg, "svg");

  const uplaodToS3 = await upload(svg, c.env);
  console.log(uplaodToS3, "uplaodToS3");



  return c.html(`
    <div>
      <h2>Generated SVG:</h2>
      ${svg}
    </div>
  `);



});

// Separate scheduler function
export default {
  /** this part manages cronjobs */
  scheduled(
    event: ScheduledEvent,
    env: Bindings,
    ctx: ExecutionContext
  ) {
    const createQuote = async () => {
      const quote = await roastthePost("generate a nice quote", "", env);
      if (quote) {
        const image = await generateImage(quote, env);
        const uplaodToS3 = await upload(image, env);
        console.log(uplaodToS3, "uplaodToS3");

        // const reply = await replyWithARoast(imageUrl: );
      }
    }
    ctx.waitUntil(createQuote());
  },
  /** this part manages regular REST */
  fetch(request: Request, env: Bindings, ctx: ExecutionContext) {
    return app.fetch(request, env, ctx);
  },
};

```

**Reply by EyalRo** [reply]
I see a couple of things that might be causing the issue:

There's a mix of ts and js; you might get better results using the js version of the export function (below).
```javascript
export default {
  scheduled(event, env, ctx) {
    const createQuote = async () => {
    ...
    }
    ctx.waitUntil(createQuote());
  },
  fetch(request, env, ctx) {
    return app.fetch(request, env, ctx)
  }
}
```
Also, be aware that if something is wrong with the `ctx.waitUntil`, you might call the function but never finish running it because the lambda is shutting down.

**Reply by serk7** [reply]
Im not sure, but 
```(javascript)
export default {
  scheduled(event, env, ctx) {
    const createQuote = async () => {
    ...
    }
    ctx.waitUntil(createQuote());
  },
  fetch(request, env, ctx) {
    return app.fetch(request, env, ctx)
  }
}
```

is the same as:

```(javascript)
export default {
  scheduled(event, env, ctx) {
    const createQuote = async () => {
    ...
    }
    ctx.waitUntil(createQuote());
  },
  fetch: app.fetch
  }
}
```
Isn't simplier?

**Comment by AustinZhu** [comment]
Check this out: https://github.com/AustinZhu/kuron

I have made a lightweight cron package for Cloudflare Cron Triggers, the APIs are designed to look familiar to Hono users.
It's type-safe, with middleware support, and you can declare multiple cron jobs!

```typescript
import { Hono } from 'hono';
import { Cron } from 'kuron';

const app = new Hono<Env>()
  .get('/health', (c) => c.text('OK'));

const cron = new Cron<Env>()
  .schedule('0 0 * * *', async (c) => {
    console.log('Daily midnight job');
  })
  .schedule('0 12 * * *', async (c) => {
    console.log('Daily noon job');
  })
  .schedule('0 0 * * 0', async (c) => {
    console.log('Weekly Sunday job');
  });

export default {
  fetch: app.fetch,
  scheduled: cron.scheduled,
};
```
